examples/spmd_mnist_classifier_fromscratch.py

```diff
@@ -72,57 +74,71 @@ def accuracy(params, batch):
+
   num_complete_batches, leftover = divmod(num_train, batch_size)
   num_batches = num_complete_batches + bool(leftover)
 
-  # For this manual SPMD example, we get the number of devices (e.g. GPUs or
-  # TPU cores) that we're using, and use it to reshape data minibatches.
-  num_devices = jax.device_count()
+  devices = np.array(jax.devices())
+  mesh = jax.make_mesh((jax.device_count(),), ("batch",))
+
+  replicated_sharding = NamedSharding(mesh, P())
+  data_sharding = NamedSharding(mesh, P("batch"))
+
   def data_stream():
     rng = npr.RandomState(0)
     while True:
       perm = rng.permutation(num_train)
       for i in range(num_batches):
-        batch_idx = perm[i * batch_size:(i + 1) * batch_size]
-        images, labels = train_images[batch_idx], train_labels[batch_idx]
-        # For this SPMD example, we reshape the data batch dimension into two
-        # batch dimensions, one of which is mapped over parallel devices.
-        batch_size_per_device, ragged = divmod(images.shape[0], num_devices)
-        if ragged:
-          msg = "batch size must be divisible by device count, got {} and {}."
-          raise ValueError(msg.format(batch_size, num_devices))
-        shape_prefix = (num_devices, batch_size_per_device)
-        images = images.reshape(shape_prefix + images.shape[1:])
-        labels = labels.reshape(shape_prefix + labels.shape[1:])
+        batch_idx = perm[i * batch_size : (i + 1) * batch_size]
+        images_np, labels_np = train_images[batch_idx], train_labels[batch_idx]
+
+        current_batch_size = images_np.shape[0]
+        if current_batch_size < batch_size:
+          pad_len = batch_size - current_batch_size
+          images_np = np.concatenate([images_np, images_np[:pad_len]], axis=0)
+          labels_np = np.concatenate([labels_np, labels_np[:pad_len]], axis=0)
+
+        images = jax.device_put(images_np, data_sharding)
+        labels = jax.device_put(labels_np, data_sharding)
         yield images, labels
+
   batches = data_stream()
 
-  @partial(pmap, axis_name='batch')
-  def spmd_update(params, batch):
-    grads = grad(loss)(params, batch)
-    # We compute the total gradients, summing across the device-mapped axis,
-    # using the `lax.psum` SPMD primitive, which does a fast all-reduce-sum.
-    grads = [(lax.psum(dw, 'batch'), lax.psum(db, 'batch')) for dw, db in grads]
-    return [(w - step_size * dw, b - step_size * db)
-            for (w, b), (dw, db) in zip(params, grads)]
-
-  # We replicate the parameters so that the constituent arrays have a leading
-  # dimension of size equal to the number of devices we're pmapping over.
-  init_params = init_random_params(param_scale, layer_sizes)
-  replicate_array = lambda x: np.broadcast_to(x, (num_devices,) + x.shape)
-  replicated_params = tree_map(replicate_array, init_params)
+  params = init_random_params(param_scale, layer_sizes)
+
+  param_shardings = tree_map(lambda x: replicated_sharding, params)
```

Comment:
Just a nit that you don't need `param_shardings = tree_map(lambda x: replicated_sharding, params)` here. 

Just use replicated_sharding from above and that should be enough.

Reply:
But do this in the follow up change!